qstack/fields/dori.py

```diff
@@ -0,0 +1,399 @@
+            # We tried to make compute_rho() write directly to these but didn't gain a lot.
+            rho[s] = rho_i
+            drho_dr[:,s] = drho_dr_i
+            d2rho_dr2[:,:,s] = d2rho_dr2_i
+        dori = compute_dori(rho, drho_dr, d2rho_dr2, eps=eps)
+        s2rho = compute_s2rho(rho, d2rho_dr2, eps=eps)
+        return dori, rho, s2rho
+
+    elif alg in [*accumulate('numerical')]:
+        dori = np.zeros_like(rho)
+        for i in grid_chunks:
+            s = np.s_[i:i+dgrid]
+            dori_i, rho_i = compute_dori_num(mol, coords[s], dm=dm, c=c, eps=eps, dx=dx)
+            dori[s] = dori_i
+            rho[s] = rho_i
+        return dori, rho, None
+
+
+def dori(mol, dm=None, c=None,
+         eps=1e-4, alg='analytical',
+         grid_type='dft',
+         grid_level=1,
+         nx=80, ny=80, nz=80, resolution=RESOLUTION, margin=BOX_MARGIN,
+         cubename=None,
+         dx=1e-4, mem=1, progress=False):
+    """Compute DORI
+
+    Args:
+        mol : an instance of :class:`Mole`.
+    Kwargs:
+        dm : 2D array of (nao,nao)
+            Density matrix (confilicts with c)
+        c : 1D array of (nao)
+            Density fitting coefficients (confilicts with dm)
+        eps : float
+            density threshold for DORI
+        alg : str
+            [a]nalytical or [n]umerical computation
+        grid_type : str
+            Type of grid, 'dft' for a DFT grid and 'cube' for a cubic grid.
+        grid_level : int
+            For a DFT grid, the grid level.
+        nx, ny, nz : int
+            For a cubic grid,
+            the number of grid point divisions in x, y, z directions.
+            Conflicts to keyword resolution.
+        resolution: float
+            For a cubic grid,
+            the resolution of the mesh grid in the cube box.
+            Conflicts to keywords nx, ny, nz.
+        cubename : str
+            For a cubic grid,
+            name for the cube files to save the results to.
+        mem : float
+              max. memory (GiB) that can be allocated to compute
+              the AO and their derivatives
+        dx : float
+            Step (in Bohr) to take the numerical derivatives
+        progress : bool
+            if print a progress bar
+
+    Returns:
+        Tuple of:
+            1D array of (ngrids) --- computed DORI
+            1D array of (ngrids) --- electron density
+            1D array of (ngrids) --- electron density * sgn(second eigenvalue of d^2rho/dr^2)
+                                     if density>=eps else 0 (only with alg='analytical').
+            2D array of (ngrids,3) --- grid coordinates
+            1D array of (ngrids) --- grid weights
+
+    """
+
+    if grid_type=='dft':
+        grid = make_grid_for_rho(mol, grid_level=grid_level)
+        weights = grid.weights
+        coords  = grid.coords
+    elif grid_type=='cube':
+        grid = Cube(mol, nx, ny, nz, resolution, margin)
+        weights = np.ones(grid.get_ngrids())
```

Comment:
are you sure this integrates correctly?
(we should have `np.allclose(n_electrons, weights@rho)`)